File: Bayesian Inference Tools/Adaptive BUS/aBUS_SuS_MATLAB/aBUS_SuS_MATLAB/main_example_1_1D.m
%% example 1: 1D posterior
%{
---------------------------------------------------------------------------
By:                Date:          Topic:
Fong-Lin Wu        July.2019      
---------------------------------------------------------------------------
Current version 2021-03
* use of log-evidence logcE
---------------------------------------------------------------------------
References:
1."Bayesian inference with subset simulation: strategies and improvements"
   Wolfgang Betz et al.
   Computer Methods in Applied Mechanics and Engineering 331 (2018) 72-93.
---------------------------------------------------------------------------
%}
clear; clc; close all; 

% add path to ERADist and ERANataf classes
% Source: https://www.bgu.tum.de/era/software/eradist/
% addpath('../../../ERA_Dist/ERADistNataf_MATLAB/')

%% initial data
d = 1;      % number of dimensions (number of uncertain parameters)

%% prior PDF for theta
prior_pdf = ERADist('normal','PAR',[0,1]);

%% likelihood PDF (mixture of gaussians in 2D)
s = RandStream.create('mt19937ar','seed',12001);   % fixed the seed to get the same data points
RandStream.setGlobalStream(s);

mu             = 5;
sigma          = 0.2;
likelihood     = @(theta) normpdf(theta,mu,sigma);
log_likelihood = @(theta) log(likelihood(theta)+realmin);   % realmin to avoid Inf values in log(0)

%% aBUS-SuS
N  = 3e3;        % number of samples per level
p0 = 0.1;        % probability of each subset

fprintf('\naBUS with SUBSET SIMULATION: \n\n');
[h, samplesU, samplesX, logcE, c, sigma] = aBUS_SuS(N, p0, log_likelihood, prior_pdf);

%% extract the samples
nsub = length(h);
u1p  = cell(nsub,1); u0p = cell(nsub,1);
x1p  = cell(nsub,1); pp  = cell(nsub,1);
for i = 1:nsub
   % samples in standard
   u1p{i} = samplesU{i}(:,1);              
   u0p{i} = samplesU{i}(:,2); 
   % samples in physical
   x1p{i} = samplesX{i}(:,1);   
   pp{i}  = samplesX{i}(:,2); 
end

%% reference and aBUS solutions
mu_exact    = 4.81;
sigma_exact = 0.196;
cE_exact    = 2.36e-6;

% show results
fprintf('\nExact model evidence = %g', cE_exact);
fprintf('\nModel evidence BUS-SuS = %g\n', exp(logcE));
fprintf('\nExact posterior mean = %g', mu_exact);
fprintf('\nMean value of samples = %g\n', mean(x1p{end}));
fprintf('\nExact posterior std = %g', sigma_exact);
fprintf('\nStd of samples = %g\n\n', std(x1p{end}));

File: Bayesian Inference Tools/Adaptive BUS/aBUS_SuS_Python/main_example_1_1D.py
"""
---------------------------------------------------------------------------
Example 1: 1D posterior
---------------------------------------------------------------------------
Created by:
Fong-Lin Wu
Felipe Uribe
Luca Sardi

Engineering Risk Analysis Group
Technische Universitat Munchen
www.bgu.tum.de/era
Contact: Antonios Kamariotis (antonis.kamariotis@tum.de)
---------------------------------------------------------------------------
Last Version 2021-03
* Adaptation to new ERANataf class
---------------------------------------------------------------------------
References:
1."Bayesian inference with subset simulation: strategies and improvements"
   Wolfgang Betz et al.
   Computer Methods in Applied Mechanics and Engineering 331 (2018) 72-93.
---------------------------------------------------------------------------
"""

#=================================================================
# import libraries
import numpy as np
import scipy as sp
from scipy.stats import multivariate_normal as mvn
import matplotlib.pylab as plt

# Make sure ERADist, ERANataf classes are in the path
# https://www.bgu.tum.de/era/software/eradist/
from ERADist import ERADist
from ERANataf import ERANataf
from aBUS_SuS import aBUS_SuS
plt.close('all')

#=================================================================
# define the prior
d           = 1        # number of dimensions (uncertain parameters)
prior_pdf_1 = ERADist('normal','PAR',[0,1])
prior_pdf   = [prior_pdf_1]

# correlation matrix
# R = np.eye(d)   # independent case

# object with distribution information
# prior_pdf = ERANataf(prior_pdf,R)

#=================================================================
# define likelihood
mu_obs    = 5
sigma_obs = 0.2
#
realmin        = np.finfo(np.double).tiny  # to prevent log(0)
likelihood     = lambda theta: mvn.pdf(theta, mu_obs, sigma_obs**2)
log_likelihood = lambda theta: np.log(likelihood(theta) + realmin)

#=================================================================
# aBUS-SuS step
N  = int(3e3)    # number of samples per level
p0 = 0.1         # probability of each subset

print('\naBUS with SUBSET SIMULATION: \n')
h, samplesU, samplesX, logcE, c, sigma = aBUS_SuS(N, p0, log_likelihood, prior_pdf)

#=================================================================
# organize samples and show results
nsub    = len(h.flatten())  # number of levels
u1p,u0p = list(),list()
x1p,pp  = list(),list()
#
for i in range(nsub):
   # samples in standard
   u1p.append(samplesU['total'][i][:,0])
   u0p.append(samplesU['total'][i][:,1])
   # samples in physical
   x1p.append(samplesX[i][:,0])
   pp.append(samplesX[i][:,1])

#=================================================================
# reference and BUS solutions
mu_exact    = 4.81
sigma_exact = 0.196
cE_exact    = 2.36e-6

# show results
print('\nExact model evidence = ', cE_exact)
print('Model evidence aBUS-SuS = ', np.exp(logcE))
print('\nExact posterior mean = ', mu_exact)
print('Mean value of samples = ', np.mean(x1p[-1]))
print('\nExact posterior std = ', sigma_exact)
print('Std of samples = ', np.std(x1p[-1]),'\n')
